File: MATLAB/Gallinaro.m
% Gallinaro et al. (2022) structural-plasticity mean-field figure.
% Translates the released Mathematica notebook time-series-fig4.nb from:
% https://github.com/juliavg/engram_structural_plasticity
%
% The script solves the deterministic mean-field ODE system used for
% Gallinaro et al., Fig. 4: population rates, calcium traces, and the 2-by-2
% excitatory connectivity matrix. Stochastic NEST connectivity samples
% distributed with the authors' public code are loaded from
% gallinaro_nest_connectivity_data.mat and overlaid as dots in panel C.

clear; clc; close all;

% --- Colour Palette ---
c_Rate    = [0.2 0.2 0.2];
c_Input   = [0.6 0.6 0.6];
c_Trace   = [0.3 0.5 0.3];
c_Facil   = [0.8 0.3 0.3];
c_Depress = [0.2 0.4 0.8];
c_Memory  = [0.6 0.4 0.8];
c_Ax      = [0.15 0.15 0.15];

%% 1. Parameters from Gallinaro et al. Fig. 4 notebook
P.J           = 0.1;
P.tau_m       = 0.02;
P.tau_r       = 0.002;
P.theta       = 20.0;
P.v_reset     = 10.0;
P.g           = 8.0;
P.epsilon     = 0.1;
P.n_e1        = 1000.0;
P.n_e2        = 9000.0;
P.n_x         = 10000.0;
P.nu_ext      = 15.0;
P.n_i         = 2500.0;
P.tau_rate    = 0.0002;
P.tau_ca      = 10.0;
P.beta_a      = 2.0;
P.beta_d      = 2.0;
P.k_a         = 1.0;
P.k_d         = 1.0;
P.t0          = 0.0;
P.tf          = 6150.0;
P.tf_plot     = 1150.0;
P.t_on        = 500.0;
P.t_off       = 650.0;
P.target_rate = 8.0;
P.n_pop       = [P.n_e1; P.n_e2];

% Numerical primitive of exp(u^2)*erfc(-u), the Brunel/Siegert integral
% in Gallinaro Eq. 13. This accelerates the same transfer function.
P.u_grid = linspace(-8.0, 8.0, 20001)';
integrand = exp(min(P.u_grid.^2, 700.0)) .* erfc(-P.u_grid);
P.primitive = cumtrapz(P.u_grid, integrand);

%% 2. Solve mean-field ODE
% State vector:
% [phi_E1; phi_E2; W_E1E1; W_E1E2; W_E2E1; W_E2E2; r_E1; r_E2; r_I]
initial = [ones(2,1) * P.target_rate; ones(4,1) * 0.1; ones(3,1) * P.target_rate];
y0 = 0.001 * initial;

options = odeset('RelTol', 1e-5, 'AbsTol', 1e-7, 'MaxStep', 2.0);
[t, y] = ode15s(@(t,y) gallinaro_mean_field_ode(t, y, P), [P.t0 P.tf], y0, options);

phi = y(:, 1:2);
w11 = y(:, 3);
w12 = y(:, 4);
w21 = y(:, 5);
w22 = y(:, 6);
rates = y(:, 7:9);
input_e1 = zeros(size(t));
for ii = 1:numel(t)
    stim_vec = stimulus(t(ii), P);
    input_e1(ii) = stim_vec(1);
end

%% 3. NEST connectivity samples from the authors' public code repository
this_file = mfilename('fullpath');
out_dir = fileparts(this_file);
data_file = fullfile(out_dir, 'gallinaro_nest_connectivity_data.mat');
if ~exist(data_file, 'file')
    error('Could not find gallinaro_nest_connectivity_data.mat. Keep it in the same folder as this script.');
end
nest_data = load(data_file);
nest_t = nest_data.sim_steps(:) / 1000.0;
nest_c11 = nest_data.c11(:);
nest_c12 = nest_data.c12(:);
nest_c21 = nest_data.c21(:);
nest_c22 = nest_data.c22(:);

%% 4. Visualization
fig = figure('Color', 'w', 'Position', [100, 100, 700, 900]);

% --- Panel A: firing-rate dynamics and stimulation ---
subplot(3,1,1);
yyaxis right;
area(t, input_e1, 'FaceColor', c_Input, 'EdgeColor', 'none', 'FaceAlpha', 0.2);
ylabel('$s_{E_1}(t)$', 'Interpreter', 'latex', 'FontSize', 12, 'Color', c_Input);
ylim([0 0.15]);
set(gca, 'YColor', c_Input, 'FontSize', 12);

yyaxis left;
plot(t, rates(:,1), 'Color', c_Rate, 'LineWidth', 2.0); hold on;
plot(t, rates(:,2), 'Color', c_Input, 'LineWidth', 1.6);
plot([P.t0 P.tf], [P.target_rate P.target_rate], '--', 'Color', [0.55 0.55 0.55], 'LineWidth', 1.1);
ylabel('$r_Y(t)$ (Hz)', 'Interpreter', 'latex', 'FontSize', 12, 'Color', c_Rate);
ylim([0 max(max(rates(:,1:2))) * 1.18]);
xlim([P.t0 P.tf_plot]);
xticks([0 50 200 350 500 650 800 950 1100]);
yl = ylim;
plot([P.t_on P.t_on], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
plot([P.t_off P.t_off], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
ylim(yl);
legend({'$r_{E_1}(t)$','$r_{E_2}(t)$'}, ...
    'Interpreter', 'latex', 'Location', 'northeast', 'Box', 'off');
set(gca, 'YColor', c_Rate, 'FontSize', 12, 'Box', 'off', 'XTickLabel', [], 'LineWidth', 1.5);

% --- Panel B: calcium trace Eq. 14 ---
subplot(3,1,2);
plot(t, phi(:,1), 'Color', c_Trace, 'LineWidth', 2.0); hold on;
plot(t, phi(:,2), 'Color', c_Input, 'LineWidth', 1.6);
plot([P.t0 P.tf], [P.target_rate P.target_rate], '--', 'Color', [0.55 0.55 0.55], 'LineWidth', 1.1);
ylabel('$\phi_Y(t)$', 'Interpreter', 'latex', 'FontSize', 12);
ylim([0 max(max(phi)) * 1.14]);
xlim([P.t0 P.tf_plot]);
xticks([0 50 200 350 500 650 800 950 1100]);
yl = ylim;
plot([P.t_on P.t_on], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
plot([P.t_off P.t_off], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
ylim(yl);
legend({'$\phi_{E_1}(t)$','$\phi_{E_2}(t)$'}, ...
    'Interpreter', 'latex', 'Location', 'northeast', 'Box', 'off');
set(gca, 'FontSize', 12, 'Box', 'off', 'XTickLabel', [], 'LineWidth', 1.5, 'YColor', c_Ax);

% --- Panel C: connectivity matrix Eq. 15, with released NEST samples ---
subplot(3,1,3);
plot(t, w11, 'Color', c_Facil, 'LineWidth', 2.0); hold on;
plot(t, w12, 'Color', c_Depress, 'LineWidth', 2.0);
plot(t, w21, 'Color', c_Memory, 'LineWidth', 2.0);
plot(t, w22, 'Color', c_Rate, 'LineWidth', 2.0);
plot(nest_t, nest_c11, 'o', 'MarkerSize', 3.5, 'MarkerFaceColor', 'w', ...
    'MarkerEdgeColor', c_Facil, 'LineWidth', 0.9);
plot(nest_t, nest_c12, 'o', 'MarkerSize', 3.5, 'MarkerFaceColor', 'w', ...
    'MarkerEdgeColor', c_Depress, 'LineWidth', 0.9);
plot(nest_t, nest_c21, 'o', 'MarkerSize', 3.5, 'MarkerFaceColor', 'w', ...
    'MarkerEdgeColor', c_Memory, 'LineWidth', 0.9);
plot(nest_t, nest_c22, 'o', 'MarkerSize', 3.5, 'MarkerFaceColor', 'w', ...
    'MarkerEdgeColor', c_Rate, 'LineWidth', 0.9);
ylabel('$\bar{C}_{Y,Z}(t)$', 'Interpreter', 'latex', 'FontSize', 12);
xlabel('Time (s)', 'FontSize', 12);
ylim([0 max([max(w11), max(nest_c11)]) * 1.08]);
xlim([P.t0 P.tf_plot]);
xticks([0 50 200 350 500 650 800 950 1100]);
yl = ylim;
plot([P.t_on P.t_on], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
plot([P.t_off P.t_off], yl, 'Color', [0.45 0.45 0.45], 'LineWidth', 1.0);
ylim(yl);
legend({'$\bar{C}_{E_1,E_1}(t)$','$\bar{C}_{E_1,E_2}(t)$', ...
    '$\bar{C}_{E_2,E_1}(t)$','$\bar{C}_{E_2,E_2}(t)$'}, ...
    'Interpreter', 'latex', 'Location', 'northoutside', ...
    'Orientation', 'horizontal', 'Box', 'off');
set(gca, 'FontSize', 12, 'Box', 'off', 'LineWidth', 1.5, 'YColor', c_Ax);

%% Local functions
function dydt = gallinaro_mean_field_ode(t, y, P)
    phi = max(y(1:2), 0.0);
    W = max([y(3) y(4); y(5) y(6)], 1e-12);
    rates = max(y(7:9), 0.0);
    r_e = rates(1:2);
    r_i = rates(3);

    k_in = max(W * P.n_pop, 1e-12);
    k_out = max(W' * P.n_pop, 1e-12);

    dend_create = max((P.target_rate - phi) * P.k_d / P.beta_d, 0.0);
    axon_create = max((P.target_rate - phi) * P.k_a / P.beta_a, 0.0);
    dend_delete = max((phi - P.target_rate) * P.k_d / P.beta_d, 0.0);
    axon_delete = max((phi - P.target_rate) * P.k_a / P.beta_a, 0.0);

    dendritic_loss = W .* ((dend_delete ./ k_in) * ones(1,2));
    axonal_loss = W .* (ones(2,1) * (axon_delete ./ k_out)');

    free_dendrites = dend_create + sum(axonal_loss .* (ones(2,1) * P.n_pop'), 2);
    free_axons = axon_create + sum(dendritic_loss .* (P.n_pop * ones(1,2)), 1)';

    total_free_dendrites = sum(P.n_pop .* free_dendrites);
    total_free_axons = sum(P.n_pop .* free_axons);
    if total_free_dendrites > 0.0 && total_free_axons > 0.0
        pairing_rate = min(total_free_dendrites, total_free_axons);
        lambda_pair = pairing_rate / (total_free_dendrites * total_free_axons);
    else
        lambda_pair = 0.0;
    end

    dW = lambda_pair * (free_dendrites * free_axons') - dendritic_loss - axonal_loss;
    dphi = (r_e - phi) / P.tau_ca;

    stim_vec = stimulus(t, P);
    e_input = (W * (P.n_pop .* r_e)) + (1.0 + stim_vec) * P.epsilon * P.n_x * P.nu_ext;
    mu_e = P.tau_m * P.J * (e_input - P.g * P.epsilon * P.n_i * r_i);
    sigma_e = P.J * sqrt(max(P.tau_m * (e_input + P.g^2 * P.epsilon * P.n_i * r_i), 1e-12));
    target_e = brunel_rate(mu_e, sigma_e, P);

    i_input = P.epsilon * sum(P.n_pop .* r_e);
    mu_i = P.tau_m * P.J * (i_input - P.g * P.epsilon * P.n_i * r_i + P.epsilon * P.n_x * P.nu_ext);
    sigma_i = P.J * sqrt(max(P.tau_m * (i_input + P.g^2 * P.epsilon * P.n_i * r_i + P.epsilon * P.n_x * P.nu_ext), 1e-12));
    target_i = brunel_rate(mu_i, sigma_i, P);

    dr = (-rates + [target_e; target_i]) / P.tau_rate;
    dydt = [dphi; dW(1,1); dW(1,2); dW(2,1); dW(2,2); dr];
end

function rate = brunel_rate(mu, sigma, P)
    sigma = max(sigma, 1e-9);
    a = (P.v_reset - mu) ./ sigma;
    b = (P.theta - mu) ./ sigma;
    integral = primitive_value(b, P) - primitive_value(a, P);
    rate = 1.0 ./ (P.tau_r + P.tau_m * sqrt(pi) .* max(integral, 0.0));
    rate(a >= P.u_grid(end)) = 0.0;
    rate(b <= P.u_grid(1)) = 1.0 / P.tau_r;
    rate(b >= P.u_grid(end)) = 0.0;
end

function values = primitive_value(x, P)
    x_clipped = min(max(x, P.u_grid(1)), P.u_grid(end));
    values = interp1(P.u_grid, P.primitive, x_clipped, 'linear');
    mask = x < P.u_grid(1);
    if any(mask)
        values(mask) = -log((-x(mask)) / (-P.u_grid(1))) / sqrt(pi);
    end
end

function stim_vec = stimulus(t, P)
    z_on = min(max(1000.0 * (t - P.t_on), -60.0), 60.0);
    z_off = min(max(1000.0 * (P.t_off - t), -60.0), 60.0);
    gate = (1.0 / (1.0 + exp(-z_on))) * (1.0 / (1.0 + exp(-z_off)));
    stim_vec = [0.10 * gate; 0.0];
end

File: Python/Gallinaro.py
"""
Gallinaro et al. (2022) structural-plasticity mean-field figure.

This script translates the Mathematica notebook `time-series-fig4.nb` from:
https://github.com/juliavg/engram_structural_plasticity

It solves the deterministic mean-field ODE system used for Gallinaro et al.,
Fig. 4: population rates, calcium traces, and the 2-by-2 excitatory
connectivity matrix. Stochastic NEST connectivity samples distributed with
the authors' public code are loaded from `gallinaro_nest_connectivity_data.mat`
and overlaid as dots in panel C.

Dependencies: numpy, scipy, matplotlib.
"""

from pathlib import Path

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp, cumulative_trapezoid
from scipy.io import loadmat
from scipy.special import erfc, expit


# --- Colour Palette ---
c_rate = (0.2, 0.2, 0.2)
c_input = (0.6, 0.6, 0.6)
c_trace = (0.3, 0.5, 0.3)
c_facil = (0.8, 0.3, 0.3)
c_depress = (0.2, 0.4, 0.8)
c_memory = (0.6, 0.4, 0.8)
c_ax = (0.15, 0.15, 0.15)


def build_parameters():
    """Parameters from Gallinaro et al. Fig. 4 notebook."""
    p = {
        "J": 0.1,
        "tau_m": 0.02,
        "tau_r": 0.002,
        "theta": 20.0,
        "v_reset": 10.0,
        "g": 8.0,
        "epsilon": 0.1,
        "n_e1": 1000.0,
        "n_e2": 9000.0,
        "n_x": 10000.0,
        "nu_ext": 15.0,
        "n_i": 2500.0,
        "tau_rate": 0.0002,
        "tau_ca": 10.0,
        "beta_a": 2.0,
        "beta_d": 2.0,
        "k_a": 1.0,
        "k_d": 1.0,
        "t0": 0.0,
        "tf": 6150.0,
        "tf_plot": 1150.0,
        "t_on": 500.0,
        "t_off": 650.0,
        "target_rate": 8.0,
    }
    p["n_pop"] = np.array([p["n_e1"], p["n_e2"]], dtype=float)
    return p


def build_transfer_primitive():
    """
    Tabulate the primitive of exp(u^2) * erfc(-u), the Brunel/Siegert
    integral used in Gallinaro Eq. 13. This is a numerical acceleration of
    the same transfer function, not a replacement dynamics.
    """
    grid = np.linspace(-8.0, 8.0, 20001)
    integrand = np.exp(np.clip(grid * grid, None, 700.0)) * erfc(-grid)
    primitive = np.r_[0.0, cumulative_trapezoid(integrand, grid)]
    return grid, primitive


def primitive_value(x, grid, primitive):
    x = np.asarray(x, dtype=float)
    x_clipped = np.clip(x, grid[0], grid[-1])
    values = np.interp(x_clipped, grid, primitive)

    # Asymptotic correction for the negative tail of exp(u^2) * erfc(-u).
    # This matters only during the very early high-input growth transient.
    mask = x < grid[0]
    if np.any(mask):
        values = np.array(values, copy=True)
        values[mask] = -np.log((-x[mask]) / (-grid[0])) / np.sqrt(np.pi)
    return values


def brunel_rate(mu, sigma, p, grid, primitive):
    """Stationary LIF firing rate f(mu, sigma) used in Gallinaro Eq. 13."""
    sigma = np.maximum(sigma, 1e-9)
    a = (p["v_reset"] - mu) / sigma
    b = (p["theta"] - mu) / sigma

    integral = primitive_value(b, grid, primitive) - primitive_value(a, grid, primitive)
    rate = 1.0 / (p["tau_r"] + p["tau_m"] * np.sqrt(np.pi) * np.maximum(integral, 0.0))

    rate = np.where(a >= grid[-1], 0.0, rate)
    rate = np.where(b <= grid[0], 1.0 / p["tau_r"], rate)
    rate = np.where(b >= grid[-1], 0.0, rate)
    return rate


def stimulus(t, p):
    """Gallinaro Fig. 4 stimulation protocol: E1 receives +10% input."""
    gate = expit(1000.0 * (t - p["t_on"])) * expit(1000.0 * (p["t_off"] - t))
    if np.ndim(gate) == 0:
        return np.array([0.10 * gate, 0.0])
    return np.vstack((0.10 * gate, np.zeros_like(gate)))


def mean_field_rhs(t, y, p, grid, primitive):
    """
    State vector:
      phi_E1, phi_E2,
      W_E1E1, W_E1E2, W_E2E1, W_E2E2,
      r_E1, r_E2, r_I.
    """
    phi = np.maximum(y[0:2], 0.0)
    w = np.maximum(y[2:6].reshape(2, 2), 1e-12)
    rates = np.maximum(y[6:9], 0.0)
    r_e = rates[:2]
    r_i = rates[2]
    n_pop = p["n_pop"]

    # Eq. 15 ingredients: average in-degree and out-degree.
    k_in = np.maximum(w @ n_pop, 1e-12)
    k_out = np.maximum(w.T @ n_pop, 1e-12)

    # Eq. 15 is written here with positive deletion magnitudes and explicit
    # subtraction, matching the released Mathematica implementation.
    dend_create = np.maximum((p["target_rate"] - phi) * p["k_d"] / p["beta_d"], 0.0)
    axon_create = np.maximum((p["target_rate"] - phi) * p["k_a"] / p["beta_a"], 0.0)
    dend_delete = np.maximum((phi - p["target_rate"]) * p["k_d"] / p["beta_d"], 0.0)
    axon_delete = np.maximum((phi - p["target_rate"]) * p["k_a"] / p["beta_a"], 0.0)

    dendritic_loss = w * (dend_delete / k_in)[:, None]
    axonal_loss = w * (axon_delete / k_out)[None, :]

    free_dendrites = dend_create + (axonal_loss * n_pop[None, :]).sum(axis=1)
    free_axons = axon_create + (dendritic_loss * n_pop[:, None]).sum(axis=0)

    total_free_dendrites = (n_pop * free_dendrites).sum()
    total_free_axons = (n_pop * free_axons).sum()
    if total_free_dendrites > 0.0 and total_free_axons > 0.0:
        pairing_rate = min(total_free_dendrites, total_free_axons)
        lambda_pair = pairing_rate / (total_free_dendrites * total_free_axons)
    else:
        lambda_pair = 0.0

    dw_dt = lambda_pair * np.outer(free_dendrites, free_axons) - dendritic_loss - axonal_loss

    # Eq. 14: calcium/activity trace.
    dphi_dt = (r_e - phi) / p["tau_ca"]

    # Eq. 13: Wilson-Cowan relaxation to the stationary LIF transfer.
    e_input = (w @ (n_pop * r_e)) + (1.0 + stimulus(t, p)) * p["epsilon"] * p["n_x"] * p["nu_ext"]
    mu_e = p["tau_m"] * p["J"] * (e_input - p["g"] * p["epsilon"] * p["n_i"] * r_i)
    sigma_e = p["J"] * np.sqrt(np.maximum(p["tau_m"] * (e_input + p["g"] ** 2 * p["epsilon"] * p["n_i"] * r_i), 1e-12))
    target_e = brunel_rate(mu_e, sigma_e, p, grid, primitive)

    i_input = p["epsilon"] * np.sum(n_pop * r_e)
    mu_i = p["tau_m"] * p["J"] * (i_input - p["g"] * p["epsilon"] * p["n_i"] * r_i + p["epsilon"] * p["n_x"] * p["nu_ext"])
    sigma_i = p["J"] * np.sqrt(
        max(p["tau_m"] * (i_input + p["g"] ** 2 * p["epsilon"] * p["n_i"] * r_i + p["epsilon"] * p["n_x"] * p["nu_ext"]), 1e-12)
    )
    target_i = brunel_rate(mu_i, sigma_i, p, grid, primitive)
    dr_dt = (-rates + np.r_[target_e, target_i]) / p["tau_rate"]

    return np.r_[dphi_dt, dw_dt.ravel(), dr_dt]


def solve_mean_field():
    p = build_parameters()
    grid, primitive = build_transfer_primitive()

    initial = np.r_[
        np.ones(2) * p["target_rate"],
        np.ones(4) * 0.1,
        np.ones(3) * p["target_rate"],
    ]
    y0 = 0.001 * initial

    t_eval = np.linspace(p["t0"], p["tf"], 2500)
    sol = solve_ivp(
        lambda t, y: mean_field_rhs(t, y, p, grid, primitive),
        (p["t0"], p["tf"]),
        y0,
        method="LSODA",
        t_eval=t_eval,
        rtol=1e-5,
        atol=1e-7,
        max_step=2.0,
    )
    if not sol.success:
        raise RuntimeError(sol.message)
    return p, sol.t, sol.y


def load_nest_connectivity():
    data_path = Path(__file__).resolve().parent / "gallinaro_nest_connectivity_data.mat"
    if not data_path.exists():
        raise FileNotFoundError(
            f"Could not find {data_path.name}. Keep it in the same folder as this script."
        )
    mat = loadmat(data_path, squeeze_me=True)
    time_s = np.asarray(mat["sim_steps"], dtype=float).ravel() / 1000.0
    data = {name: np.asarray(mat[name], dtype=float).ravel() for name in ["c11", "c12", "c21", "c22"]}
    return time_s, data


def make_figure():
    p, t, y = solve_mean_field()
    phi = y[0:2, :]
    w = y[2:6, :].T.reshape(-1, 2, 2)
    rates = y[6:9, :]
    input_e1 = stimulus(t, p)[0]
    nest_t, nest = load_nest_connectivity()

    plt.rcParams.update({
        "font.family": "Arial",
        "font.size": 12,
        "axes.linewidth": 1.5,
        "axes.spines.top": False,
        "axes.spines.right": False,
        "xtick.major.width": 1.2,
        "ytick.major.width": 1.2,
        "legend.frameon": False,
    })

    fig, axes = plt.subplots(3, 1, figsize=(7, 9), sharex=True)
    fig.patch.set_facecolor("white")

    # --- Panel A: firing-rate dynamics and stimulation ---
    ax = axes[0]
    ax2 = ax.twinx()
    ax2.area = ax2.fill_between(t, 0.0, input_e1, color=c_input, alpha=0.2, linewidth=0)
    ax2.set_ylim(0.0, 0.15)
    ax2.set_ylabel(r"$s_{E_1}(t)$", color=c_input)
    ax2.tick_params(axis="y", colors=c_input)
    ax2.spines["right"].set_visible(True)
    ax2.spines["right"].set_linewidth(1.5)
    ax2.spines["right"].set_color(c_input)
    ax2.spines["top"].set_visible(False)

    ax.plot(t, rates[0], color=c_rate, linewidth=2.0, label=r"$r_{E_1}(t)$")
    ax.plot(t, rates[1], color=c_input, linewidth=1.6, label=r"$r_{E_2}(t)$")
    ax.axhline(p["target_rate"], color=c_rate, linestyle="--", linewidth=1.1, alpha=0.45)
    ax.set_ylabel(r"$r_Y(t)$ (Hz)", color=c_rate)
    ax.tick_params(axis="y", colors=c_rate)
    ax.set_ylim(0.0, max(rates[0].max(), rates[1].max()) * 1.18)
    ax.legend(loc="upper right", handlelength=1.8)

    # --- Panel B: calcium trace Eq. 14 ---
    ax = axes[1]
    ax.plot(t, phi[0], color=c_trace, linewidth=2.0, label=r"$\phi_{E_1}(t)$")
    ax.plot(t, phi[1], color=c_input, linewidth=1.6, label=r"$\phi_{E_2}(t)$")
    ax.axhline(p["target_rate"], color=c_rate, linestyle="--", linewidth=1.1, alpha=0.45)
    ax.set_ylabel(r"$\phi_Y(t)$")
    ax.set_ylim(0.0, max(phi[0].max(), phi[1].max()) * 1.14)
    ax.legend(loc="upper right", handlelength=1.8)

    # --- Panel C: connectivity matrix Eq. 15, with released NEST samples ---
    ax = axes[2]
    labels = [
        (0, 0, r"$\bar{C}_{E_1,E_1}(t)$", c_facil, "c11"),
        (0, 1, r"$\bar{C}_{E_1,E_2}(t)$", c_depress, "c12"),
        (1, 0, r"$\bar{C}_{E_2,E_1}(t)$", c_memory, "c21"),
        (1, 1, r"$\bar{C}_{E_2,E_2}(t)$", c_rate, "c22"),
    ]
    for row, col, label, color, data_key in labels:
        ax.plot(t, w[:, row, col], color=color, linewidth=2.0, label=label)
        ax.plot(
            nest_t,
            nest[data_key],
            linestyle="none",
            marker="o",
            markersize=3.5,
            markerfacecolor="white",
            markeredgecolor=color,
            markeredgewidth=0.9,
            alpha=0.85,
        )
    ax.set_ylabel(r"$\bar{C}_{Y,Z}(t)$")
    ax.set_xlabel("Time (s)")
    ax.set_ylim(0.0, max(w[:, 0, 0].max(), nest["c11"].max()) * 1.08)
    ax.legend(
        loc="lower center",
        bbox_to_anchor=(0.5, 1.02),
        ncol=4,
        handlelength=1.8,
        columnspacing=0.9,
    )

    for axis in axes:
        axis.axvline(p["t_on"], color=c_ax, linewidth=1.0, alpha=0.65)
        axis.axvline(p["t_off"], color=c_ax, linewidth=1.0, alpha=0.65)
        axis.set_xlim(p["t0"], p["tf_plot"])
        axis.set_xticks([0, 50, 200, 350, 500, 650, 800, 950, 1100])
        axis.grid(False)
        axis.spines["top"].set_visible(False)
        axis.spines["right"].set_visible(False)
        axis.tick_params(labelsize=12)

    fig.tight_layout(h_pad=1.15)


if __name__ == "__main__":
    make_figure()
